src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb

```diff
@@ -262,24 +272,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 Dim array1 = DirectCast(t1, ArrayTypeSymbol)
                 Dim array2 = DirectCast(t2, ArrayTypeSymbol)
 
-                Return array1.HasSameShapeAs(array2) AndAlso
-                       array1.ElementType.IsSameTypeIgnoringCustomModifiers(array2.ElementType)
-
+                Return array1.IsSameType(array2, compareKind)
             ElseIf t1.IsAnonymousType AndAlso t2.IsAnonymousType Then
-                Return AnonymousTypeManager.EqualsIgnoringCustomModifiers(t1, t2)
+                Return AnonymousTypeManager.IsSameType(t1, t2, compareKind)
 
             ElseIf kind = SymbolKind.NamedType OrElse kind = SymbolKind.ErrorType Then
                 Dim t1IsDefinition = t1.IsDefinition
                 Dim t2IsDefinition = t2.IsDefinition
 
                 If (t1IsDefinition <> t2IsDefinition) AndAlso
+                   (compareKind And TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) = 0 AndAlso
                    Not (DirectCast(t1, NamedTypeSymbol).HasTypeArgumentsCustomModifiers OrElse DirectCast(t2, NamedTypeSymbol).HasTypeArgumentsCustomModifiers) Then
                     Return False
                 End If
 
+                If (compareKind And TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) = 0 Then
+                    Dim mod1 = DirectCast(t1, NamedTypeSymbol).TypeArgumentsCustomModifiers
+                    Dim mod2 = DirectCast(t2, NamedTypeSymbol).TypeArgumentsCustomModifiers
+
+                    If mod1.Length <> mod2.Length Then
+                        Return False
+                    End If
+
+                    If Not mod1.SequenceEqual(mod2, AddressOf AreSameCustomModifiers) Then
+                        Return False
+                    End If
+                End If
+
                 If Not (t1IsDefinition AndAlso t2IsDefinition) Then ' This is a generic instantiation case
 
-                    If t1.OriginalDefinition <> t2.OriginalDefinition Then
+                    If Not t1.OriginalDefinition.IsSameType(t2.OriginalDefinition, compareKind) Then
```

Comment:
>IsSameType [](start = 49, length = 10)

I think this is redundant recursion. Original definitions can never be different due to custom modifiers or tuple names, because they are not supposed to have any of them.